### PR TITLE
rpc: Shuffle inputs and outputs after joining psbts

### PR DESCRIPTION
**Pull request by achow101**
`joinpsbts` currently just adds the inputs and outputs in the order of that the PSBTs were provided. This makes it extremely easy to identify which outputs belong to which inputs. This PR changes that so that all of the inputs and outputs are shuffled in the joined transaction.

### CODE REVIEW

File: src/rpc/rawtransaction.cpp [hunk 1]
```diff
@@ -1604,8 +1604,29 @@ UniValue joinpsbts(const JSONRPCRequest& request)
...
 
+    // Generate list of shuffled indices
+    std::vector<int> input_indices(merged_psbt.inputs.size());
+    std::iota(input_indices.begin(), input_indices.end(), 0);
+    Shuffle(input_indices.begin(), input_indices.end(), FastRandomContext());
+    std::vector<int> output_indices(merged_psbt.outputs.size());
+    std::iota(output_indices.begin(), output_indices.end(), 0);
+    Shuffle(output_indices.begin(), output_indices.end(), FastRandomContext());
+
+    // Shuffle the inputs
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: Unnecessary comment? Or maybe move it to L1607 like `Shuffle inputs and outputs of the merged PSBT`.

**Review by promag — COMMENTED** [review]
Concept ACK.

**Comment by laanwj** [comment]
Concept ACK, not sure whether there's reasons someone would want to disable this, but in any case I think privacy-by-default behavior is good

**Comment by practicalswift** [comment]
Concept ACK

File: src/rpc/rawtransaction.cpp (hunk 1 above)
Comment (achow101) [reply src/rpc/rawtransaction.cpp]: Removed the comment and updated the one above.

**Comment by fanquake** [comment]
Concept ACK - Can we get some sort of simple test case to ensure that there's some shuffling going on.

> but in any case I think privacy-by-default behavior is good

Definitely agree.

I guess this should also be backported given the patch is fairly simple and it's a privacy improvement?

**Comment by achow101** [comment]
I've added a test. It runs joinpsbts multiple times and checks whether at least one of the resulting psbts are different.

File: test/functional/rpc_psbt.py [hunk 2]
```diff
@@ -370,6 +370,13 @@ def test_psbt_input_keys(psbt_input, keys):
         joined_decoded = self.nodes[0].decodepsbt(joined)
         assert len(joined_decoded['inputs']) == 4 and len(joined_decoded['outputs']) == 2 and "final_scriptwitness" not in joined_decoded['inputs'][3] and "final_scriptSig" not in joined_decoded['inputs'][3]
 
+        # Check that joining shuffles the inputs and outputs
+        shuffled = False
+        for i in range(0, 10):
+            shuffled_joined = self.nodes[0].joinpsbts([psbt, psbt2])
+            shuffled |= joined != shuffled_joined
```
Comment (promag) [inline test/functional/rpc_psbt.py]: nit, could break if shuffled.

File: test/functional/rpc_psbt.py [hunk 3]
```diff
@@ -370,6 +370,13 @@ def test_psbt_input_keys(psbt_input, keys):
         joined_decoded = self.nodes[0].decodepsbt(joined)
         assert len(joined_decoded['inputs']) == 4 and len(joined_decoded['outputs']) == 2 and "final_scriptwitness" not in joined_decoded['inputs'][3] and "final_scriptSig" not in joined_decoded['inputs'][3]
 
+        # Check that joining shuffles the inputs and outputs
```
Comment (promag) [inline test/functional/rpc_psbt.py]: nit, could improve comment like "10 attempts should be enough to get a shuffled join".

File: src/rpc/rawtransaction.cpp [hunk 4]
```diff
@@ -1604,8 +1604,28 @@ UniValue joinpsbts(const JSONRPCRequest& request)
         merged_psbt.unknown.insert(psbt.unknown.begin(), psbt.unknown.end());
     }
 
+    // Generate list of shuffled indices for shuffling inputs and outputs of the merged PSBT
+    std::vector<int> input_indices(merged_psbt.inputs.size());
+    std::iota(input_indices.begin(), input_indices.end(), 0);
+    Shuffle(input_indices.begin(), input_indices.end(), FastRandomContext());
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: Should `#include <random.h>`

**Review by promag — COMMENTED** [review]
ACK, but could add a release note regarding the behavior change and also improve help text.

File: test/functional/rpc_psbt.py (hunk 2 above)
Comment (achow101) [reply test/functional/rpc_psbt.py]: Done

File: test/functional/rpc_psbt.py (hunk 3 above)
Comment (achow101) [reply test/functional/rpc_psbt.py]: Done

File: src/rpc/rawtransaction.cpp (hunk 4 above)
Comment (achow101) [reply src/rpc/rawtransaction.cpp]: Done

**Comment by achow101** [comment]
Added a release note.

**Comment by luke-jr** [comment]
Only bugfixes should be backported, not improvements...

**Comment by JeremyRubin** [comment]
Wouldn't it be better to deterministically sort them?

**Comment by achow101** [comment]
> Wouldn't it be better to deterministically sort them?

Not at all. That gives a different privacy leak as it will indicate that `joinpsbts` has been used on that transaction.

**Comment by JeremyRubin** [comment]
Hm gotcha. I was unclear of the status of BIP-69 and related standard adoption efforts for Core

File: test/functional/rpc_psbt.py [hunk 5]
```diff
@@ -370,6 +370,16 @@ def test_psbt_input_keys(psbt_input, keys):
...
+        # 10 attempts should be enough to get a shuffled join
+        shuffled = False
```
[the remark this answers is not in the document]
Comment (promag) [reply test/functional/rpc_psbt.py]: Do you mean just replacing the line below with `while True:`? If you mean to drop `shuffled` then what would you assert after the loop?

[the remark this answers is not in the document]
Comment (promag) [reply test/functional/rpc_psbt.py]: I agree with you but it it isn't that informative. Suppose it hangs in travis, you would be clueless.

**Review by jonatack — COMMENTED** [review]
Concept ACK b1d769faf3bab38c18c7424cfa5f1aeadc832c30. Code review, built, ran tests, verified the added test fails without the change. Code looks correct; a few suggestions follow for your perusal. Is testing joinpsbt inequality in rpc_psbt.py enough? Could the manual copy+shuffling be extracted and covered by unit tests? e.g. as a member function of PartiallySignedTransaction?

File: src/rpc/rawtransaction.cpp [hunk 6]
```diff
@@ -1604,8 +1605,28 @@ UniValue joinpsbts(const JSONRPCRequest& request)
         merged_psbt.unknown.insert(psbt.unknown.begin(), psbt.unknown.end());
     }
 
+    // Generate list of shuffled indices for shuffling inputs and outputs of the merged PSBT
```
[the remark this answers is not in the document]
Comment (jonatack) [reply src/rpc/rawtransaction.cpp]: Note that the proposed change places the two Shuffle functions together. First generate the lists, then perform the shuffles.

**Comment by nopara73** [comment]
Concept ACK. Related: https://github.com/bitcoin/bitcoin/issues/12457

File: test/functional/rpc_psbt.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (achow101) [reply test/functional/rpc_psbt.py]: I think 10 is enough for now.

File: test/functional/rpc_psbt.py [hunk 7]
```diff
@@ -370,6 +370,16 @@ def test_psbt_input_keys(psbt_input, keys):
...
+        for i in range(0, 10):
```
[the remark this answers is not in the document]
Comment (achow101) [reply test/functional/rpc_psbt.py]: meh.

File: test/functional/rpc_psbt.py [hunk 8]
```diff
@@ -370,6 +370,16 @@ def test_psbt_input_keys(psbt_input, keys):
         joined_decoded = self.nodes[0].decodepsbt(joined)
         assert len(joined_decoded['inputs']) == 4 and len(joined_decoded['outputs']) == 2 and "final_scriptwitness" not in joined_decoded['inputs'][3] and "final_scriptSig" not in joined_decoded['inputs'][3]
 
+        # Check that joining shuffles the inputs and outputs
```
[the remark this answers is not in the document]
Comment (achow101) [reply test/functional/rpc_psbt.py]: meh.

File: src/rpc/rawtransaction.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (achow101) [reply src/rpc/rawtransaction.cpp]: Done

File: doc/release-notes-16512.md [hunk 9]
```diff
@@ -0,0 +1,3 @@
+RPC changes
+-----------
+The RPC `joinpsbts` will shuffle the order of the inputs and outputs of the resulting joined psbt.
```
[the remark this answers is not in the document]
Comment (achow101) [reply doc/release-notes-16512.md]: Done

File: src/rpc/rawtransaction.cpp [hunk 10]
```diff
@@ -1604,8 +1605,30 @@ UniValue joinpsbts(const JSONRPCRequest& request)
...
+    std::vector<int> input_indices(merged_psbt.inputs.size());
+    std::iota(input_indices.begin(), input_indices.end(), 0);
+    std::vector<int> output_indices(merged_psbt.outputs.size());
+    std::iota(output_indices.begin(), output_indices.end(), 0);
+
+    // Shuffle input and output indicies lists
```
Comment (jonatack) [inline src/rpc/rawtransaction.cpp]: nit (if you need to retouch this): s/indicies/indices/

File: test/functional/rpc_psbt.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (instagibbs) [reply test/functional/rpc_psbt.py]: first test failure you're getting strung up :P

**Comment by jonatack** [comment]
@achow101 what did you think of [this question / suggestion](https://github.com/bitcoin/bitcoin/pull/16512#pullrequestreview-283522359) above?

**Comment by achow101** [comment]
> Is testing joinpsbt inequality in rpc_psbt.py enough?

It should be.

> Could the manual copy+shuffling be extracted and covered by unit tests? e.g. as a member function of PartiallySignedTransaction?

Maybe later? I would like this merged for 0.19 so I do want any further changes to this to be minimal.

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/16512/commits/c0b5d9710322a614a50ab5da081558cf6a38ad2a

**Comment by jonatack** [comment]
> > Could the manual copy+shuffling be extracted and covered by unit tests? e.g. as a member function of PartiallySignedTransaction?
> 
> Maybe later? I would like this merged for 0.19 so I do want any further changes to this to be minimal.

Fair enough. I might propose it later.

**Comment by jonatack** [comment]
ACK c0b5d9710322a614a50ab5da081558cf6a38ad2a modulo suggestions for later.

**Comment by laanwj** [comment]
ACK c0b5d9710322a614a50ab5da081558cf6a38ad2a

Agree the shuffling should ideally be factored out to a function, but that can be done in a future PR

**Comment by fanquake** [comment]
Added for backport in #16617.

**Pull request merged**